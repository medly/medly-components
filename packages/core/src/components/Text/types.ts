import { FontVariants, FontWeights } from '@medly-components/theme';
import { HTMLProps } from '@medly-components/utils';

interface CommonProps extends HTMLProps<HTMLElement> {
    /** Text color */
    textColor?: string;
    /** To use uppercase text */
    uppercase?: boolean;
    /** To use line through text */
    lineThrough?: boolean;
    /** Takes full width of the parent component */
    fullWidth?: boolean;
    /** To be used to render Text as any component*/
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
}

export interface TextProps extends CommonProps {
    /** Text Variant */
    textVariant?: FontVariants;
    /** Text weight */
    textWeight?: FontWeights;
    /** Text alignment */
    textAlign?:
        | 'left'
        | 'right'
        | 'center'
        | 'justify'
        | 'justify-all'
        | 'start'
        | 'end'
        | 'match-parent'
        | 'inherit'
        | 'initial'
        | 'unset';
}
