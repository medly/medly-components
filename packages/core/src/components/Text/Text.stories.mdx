import { Text } from './Text';
import { Preview, Story, Meta, Props } from '@storybook/addon-docs/blocks';
import { Basic, VerticalAlign } from './Text.stories';

<Meta title="Core" component={Text} />

# Text

The `Text` component displays text efficiently. It supports nesting, styling, and touch handling. 

<Preview>
    <Story name="Text">{Basic()}</Story>
</Preview>

## Text with different Variants

You can use the font variant to decide the font style (`font-size`,`font-weight`,`line-height`,`letter-spacing`) as well as the `html tag`. If you use the `textWeight` prop as `Strong`|`ExtraStrong`, then both varaints `body1` and `body2` will use the `strong` html tag.

| Variant | Html Tag |
| ------- | -------- |
| h1      | h1       |
| h2      | h2       |
| h3      | h3       |
| h4      | h4       |
| h5      | h5       |
| body1   | span     |
| body2   | span     |

<Preview>
    <VerticalAlign>
        <Text textVariant="h1">Header 1 / Open Sans - Medium</Text>
        <Text textVariant="h2">Header 2 / Open Sans - Medium</Text>
        <Text textVariant="h3">Header 3 / Open Sans - Regular</Text>
        <Text textVariant="h4">Header 4 / Open Sans - Strong</Text>
        <Text textVariant="h5">HEADER 5 / Open Sans - Strong</Text>
        <Text textVariant="body1">Body 1 / Open Sans - Regular</Text>
        <Text textVariant="body2">Body 2 / Open Sans - Regular</Text>
    </VerticalAlign>
</Preview>

## Text with different fontWeight

You can use the different font weights to define the thinness or thickness of a font like mentioned below.

<Preview>
    <VerticalAlign>
        <Text textWeight="Light">Body 2 / Open Sans - Light</Text>
        <Text textWeight="Regular">Body 2 / Open Sans - Regular</Text>
        <Text textWeight="Medium">Body 2 / Open Sans - Medium</Text>
        <Text textWeight="Strong">Body 2 / Open Sans - Strong</Text>
        <Text textWeight="ExtraStrong">Body 2 / Open Sans - Extra Strong</Text>
    </VerticalAlign>
</Preview>

## Props

This component offers the following props mentioned in the table below:

<Props of={Text} />
