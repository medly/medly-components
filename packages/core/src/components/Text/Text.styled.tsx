import { defaultTheme } from '@medly-components/theme';
import { WithThemeProp } from '@medly-components/utils';
import styled, { css } from 'styled-components';
import { TextProps } from './types';

const uppercase = () => css`
    text-transform: uppercase;
`;

const lineThrough = () => css`
    text-decoration: line-through;
`;
const getTextStyle = ({ theme, ...props }: TextProps & WithThemeProp) => {
    const { textColor, textVariant, textWeight, textAlign } = props,
        { defaults, weights, variants } = theme.font,
        { fontSize, fontWeight, lineHeight, letterSpacing } = variants[textVariant || defaults.variant];

    return css`
        margin: 0;
        color: ${textColor};
        font-size: ${fontSize};
        font-weight: ${weights[textWeight || fontWeight]};
        letter-spacing: ${letterSpacing};
        line-height: ${lineHeight};
        text-align: ${textAlign};
    `;
};

export const TextStyled = styled('span')<TextProps>`
    ${getTextStyle}

    ${({ fullWidth, theme }) =>
        fullWidth &&
        css`
            display: block;
            margin: ${theme.spacing.S2} 0;
        `}

    ${props => props.uppercase && uppercase()};
    ${props => props.lineThrough && lineThrough()};
    display: ${({ as }) => (as === 'strong' || as === 'span' || as === 'a') && 'inline-block'};
`;

TextStyled.defaultProps = {
    theme: defaultTheme
};
