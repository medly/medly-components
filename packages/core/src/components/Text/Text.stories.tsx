import { FontVariants, FontWeights } from '@medly-components/theme';
import { boolean, color, select } from '@storybook/addon-knobs';
import styled from 'styled-components';
import { Text } from './Text';

const variantOptions: FontVariants[] = ['h1', 'h2', 'h3', 'h4', 'h5', 'body1', 'body2'];
const weightOptions: FontWeights[] = ['Light', 'Regular', 'Medium', 'Strong'];

export const Basic = () => (
    <Text
        fullWidth={boolean('Full Width', false)}
        uppercase={boolean('Upper Case', false)}
        lineThrough={boolean('Line through', false)}
        textColor={color('Color', '#012040')}
        textVariant={select('Text Variant', variantOptions, 'body2')}
        textWeight={select('Text Weight', weightOptions, 'Regular')}
    >
        Lorem ipsum dolor sit, amet consectetur adipisicing elit.
    </Text>
);

export const VerticalAlign = styled.div`
    width: 100%;
    display: flex;
    flex-direction: column;
    & > * {
        margin-bottom: 8px;
    }
`;
