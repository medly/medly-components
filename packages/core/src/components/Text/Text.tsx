import { WithStyle } from '@medly-components/utils';
import React, { FC, useMemo } from 'react';
import { TextStyled } from './Text.styled';
import { TextProps } from './types';

export const Text: FC<TextProps> & WithStyle = React.memo(
    React.forwardRef(({ as, textVariant, textWeight, textAlign, children, ...restProps }, ref) => {
        const componentType = useMemo(() => {
            let type: keyof JSX.IntrinsicElements | React.ComponentType<any> = (textVariant as keyof JSX.IntrinsicElements) || 'span';

            if (
                !textVariant ||
                textVariant === 'body1' ||
                textVariant == 'body2' ||
                textVariant == 'body3' ||
                textVariant == 'button1' ||
                textVariant == 'button2'
            ) {
                type = textWeight === 'Strong' || textWeight === 'ExtraStrong' ? 'strong' : 'span';
            }

            return type;
        }, [textWeight, textVariant]);
        return (
            <TextStyled {...{ textVariant, textWeight, textAlign, ...restProps }} as={as || componentType} ref={ref}>
                {children}
            </TextStyled>
        );
    })
);

Text.defaultProps = {
    lineThrough: false,
    uppercase: false,
    fullWidth: false,
    textColor: 'inherit',
    textAlign: 'initial'
};
Text.displayName = 'Text';
Text.Style = TextStyled;
