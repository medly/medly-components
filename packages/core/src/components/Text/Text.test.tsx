import { render } from '@test-utils';
import { Text } from './Text';

describe('Text Component', () => {
    it('should render span element by default', () => {
        const { container } = render(<Text />);
        expect(container).toMatchSnapshot();
    });

    it('should render strong html tag for textWeight Strong', () => {
        const { container } = render(<Text textWeight="Strong" />);
        expect(container).toMatchSnapshot();
    });

    it('should render strong html tag for textWeight ExtraStrong', () => {
        const { container } = render(<Text textWeight="ExtraStrong" />);
        expect(container).toMatchSnapshot();
    });

    it('should render with all the props', () => {
        const { container } = render(<Text fullWidth textVariant="body2" textWeight="Light" textColor="pink" uppercase lineThrough />);
        expect(container).toMatchSnapshot();
    });
});
