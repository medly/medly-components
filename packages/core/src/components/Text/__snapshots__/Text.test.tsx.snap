// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component should render span element by default 1`] = `
.c0 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 400;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

<div>
  <span
    class="c0"
  />
</div>
`;

exports[`Text Component should render strong html tag for textWeight ExtraStrong 1`] = `
.c0 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 800;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

<div>
  <strong
    class="c0"
  />
</div>
`;

exports[`Text Component should render strong html tag for textWeight Strong 1`] = `
.c0 {
  margin: 0;
  color: inherit;
  font-size: 1.4rem;
  font-weight: 700;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: inline-block;
}

<div>
  <strong
    class="c0"
  />
</div>
`;

exports[`Text Component should render with all the props 1`] = `
.c0 {
  margin: 0;
  color: pink;
  font-size: 1.4rem;
  font-weight: 300;
  -webkit-letter-spacing: 0rem;
  -moz-letter-spacing: 0rem;
  -ms-letter-spacing: 0rem;
  letter-spacing: 0rem;
  line-height: 2.2rem;
  text-align: initial;
  display: block;
  margin: 0.8rem 0;
  text-transform: uppercase;
  -webkit-text-decoration: line-through;
  text-decoration: line-through;
  display: inline-block;
}

<div>
  <span
    class="c0"
  />
</div>
`;
